catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/util/impl/SourcePoller.java

```diff
@@ -13,110 +13,429 @@
+    // throwing an Exception, the new value was already successfully cached. If an Exception is
+    // thrown, check if it is a TimeoutException. In that case, cache the valueOnTimeout. Otherwise,
+    // there is some other unexpected (thread-related) error so do not update the cache.
+    for (final Map.Entry<Future, Pair<SourceKey, Optional<T>>> futureSourceKeyAndOldValueEntry :
+        futureSourceKeyAndOptionalOldValueMap.entrySet()) {
+      final Future future = futureSourceKeyAndOldValueEntry.getKey();
+      final SourceKey sourceKey = futureSourceKeyAndOldValueEntry.getValue().getLeft();
+      final Optional<T> optionalOldValue = futureSourceKeyAndOldValueEntry.getValue().getRight();
+
+      try {
+        future.get(); // TODO Would it be safe to still add a timeout here?
+      } catch (CancellationException e) {
+        future.cancel(true);
+        LOGGER.debug(
+            "CancellationException while getting the {} for source {}. Not updating the cache",
+            valueName,
+            sourceKey,
+            e);
+      } catch (ExecutionException e) {
+        future.cancel(true);
+        if (e.getCause() instanceof TimeoutException) {
+          LOGGER.debug("Timeout occurred while getting the {} for source {}", valueName, sourceKey);
+
+          if (optionalValueOnTimeout.isPresent()) {
+            final T valueOnTimeout = optionalValueOnTimeout.get();
+            cacheNewValue(sourceKey, optionalOldValue, valueOnTimeout);
+          } else {
+            LOGGER.debug(
+                "Not updating the cache for the timeout encountered when getting the {} for source {} because the optionalValueOnTimeout is empty",
+                valueName,
+                sourceKey);
+          }
+        } else {
+          LOGGER.debug(
+              "ExecutionException while getting the {} for source {}. Not updating the cache",
+              valueName,
+              sourceKey,
+              e);
+        }
+      } catch (InterruptedException e) {
+        future.cancel(true);
+        LOGGER.debug(
+            "Interrupted while getting the {} for source {}. Not updating the cache",
+            valueName,
+            sourceKey);
+        Thread.currentThread().interrupt();
+      }
+    }
+
+    LOGGER.trace(
+        "Successfully updated the cache with the {} for each of the {} current sources",
+        valueName,
+        currentSourcesCount);
+  }
+
+  private void cacheNewValue(
+      final SourceKey sourceKey, final Optional<T> optionalOldValue, final T newValue) {
+    LOGGER.trace("Caching {}={} for source {}", valueName, newValue, sourceKey);
+    cache.put(sourceKey, newValue);
+
+    if (optionalOldValue.isPresent()) {
+      final T oldValue = optionalOldValue.get();
+      if (!java.util.Objects.equals(newValue, oldValue)) {
+        LOGGER.info(
+            "The {} for source {} was updated from {} to {}",
+            valueName,
+            sourceKey,
+            oldValue,
```

Comment:
❓ We're only using the old cached value to create better logs right? I see the appeal but I think it might make the source polling method a little easier to read if we didn't have to carry this old value all the way through it.

Reply:
That is right. It is only used for the logs. However, it is used to print INFO-level logs when the value changes, which I think is a valuable improvement from the old SourcePoller.